README.md

```diff
@@ -1,32 +1,65 @@
+   * | ArrowBlockMetadata | attr_offsets[num_col] (32) | bitmap for slots (64-bit aligned) | data (64-bit aligned)   |
+   * -----------------------------------------------------------------------------------------------------------------
+   *
+```
 
-   We hope to have our first release in 2020.
+### Creating Checkpoints
+We plan to take checkpoints on an epoch-based rule. Checkpoints will be taken in intervals by scanning through every table. For each data table, we write its content specified by the format mentioned above into a single file. Therefore, a single checkpoint would contain a series of data table files. Eventually, we will write all these data table files into a single file to avoid exploding the filesystem.
 
-4. **Will the new system still be "[self-driving](http://www.cs.cmu.edu/~pavlo/blog/2018/04/what-is-a-self-driving-database-management-system.html)"?**
+### Recovery
+Recovery consists of three steps:
+The first step is to recover the catalog. This is a tricky step as the catalog information is also contained in the data table files. We are still developing ideas for recovering the catalog. Currently, we use the existing log recovery for restoring the catalog.
+The second step is to recover all the data tables. The files are interpreted and dumped directly into the DataTables. For var-length datas, we reconstruct new pointers to the var-length content and replace the deprecated pointers.
+The third step is to recover the index. One solution would be to re-insert all the tuples into the index one by one and rebuild the index from the start. The other solution would be to store the primary index as well, so we can load the index directly. This approach, however, requires careful handling with pointers.
 
-   Yes, our goal is to have this new system to support autonomous operation and optimization as a first-class design principle.
 
-5. **Will the new system still be PostgreSQL compatiable?**
-   
-   Yes. The DBMS supports the PostgresSQL network protocol (both simple and extended) and emulates PostgresSQL's catalog layout. We are working on support for PL/pgSQL UDFs.
+## Design Rationale
+On a high level, we seperate the recovery of catalog tables from recovery from data tables. For catalog information and catalog table data and indices, we make use of the existing write-ahead logs to recover those portions. For all other user created data tables, we recover the content of the data tables with our checkpoint mechanism. This approach simplifies the implementation of checkpointing to stay in the scope of user-defined data-table. It also provides us with the assumption that catalog tables are fully recovered before we recover from the checkpoint. We believe this approach will still take the advantage of using checkpoint, since in a larger database the majority of data that requires recovery will lie within user defined data tables.
```

Comment:
I think you have a serious flaw in the value of a checkpoint that excludes the catalog.

Currently, both DDL and DML are logged in a single stream.  This means in order to reconstruct the catalog you have to scan the entire log (with disk IO being the expensive part of log replay).  Additionally, your assumption further assumes that no DDL has occurred since the most recent checkpoint.  If it has, then either you must either retain all logs forever (defeats the main advantage of checkpointing) or risk significant data loss if a crash occurs after DDL but before completion of next checkpoint.